webapp/shop/advantage/views.py

```diff
@@ -34,6 +35,121 @@
 )
 
 
+@shop_decorator(area="advantage", permission="user", response="html")
+def get_activate_view(ua_contracts_api, advantage_mapper, **kwargs):
+    account = None
+    user = user_info(flask.session)
+    name = user["fullname"]
+    try:
+        account = advantage_mapper.get_purchase_account("canonical-ua")
+        if account.type == "paid":
+            return flask.render_template(
+                "pro/activate.html",
+                name=name,
+                needs_paid_account_created=False,
+            )
+    except UAContractsUserHasNoAccount:
+        return flask.render_template(
+            "pro/activate.html", name=name, needs_paid_account_created=True
+        )
+    except AccessForbiddenError:
+        return flask.render_template("account/forbidden.html")
+    else:
+        flask.render_template("account/forbidden.html")
+
+
+@shop_decorator(area="advantage", permission="user", response="html")
+def pro_activate_activation_key(ua_contracts_api, advantage_mapper, **kwargs):
+    marketplace = kwargs.get("marketplace", "canonical-ua")
+    try:
+        if (
+            flask.request.method == "POST"
+            and flask.request.form["submit"] == "buying-for-submit"
+        ):
+            return activate_page_buying_for_submit(
+                marketplace, ua_contracts_api
+            )
+
+        if (
+            flask.request.method == "POST"
+            and flask.request.form["submit"] == "activate-submit"
+        ):
+            return activate_page_activate_submit(
+                advantage_mapper, ua_contracts_api
+            )
+    except UAContractsUserHasNoAccount:
+        pass
+    except AccessForbiddenError:
+        return flask.render_template("account/forbidden.html")
+
+
+def activate_page_activate_submit(advantage_mapper, ua_contracts_api):
+    account = advantage_mapper.get_purchase_account("canonical-ua")
+    name = account.name
+    account_id = account.id
+    pro_activation_key = flask.request.form.get("pro-activation-key")
+    if not is_pro_activation_key_valid(pro_activation_key):
```

Comment:
Would the API not return an error if the keys is not in a valid format anyway?

Reply:
Actually, it was my suggestion (so me to blame if this is not correct 😅), to reduce the number of inherently incorrect requests coming to the ua-contracts API.

I do not have any strong opinion here, feel free to overwrite.

Reply:
This block is removed and now this validation is done in the front-end so it will only make a request to the API if the key starts with K and is 23 long.